unstructured/embed/embedder/open_ai.py

```diff
@@ -0,0 +1,155 @@
+    """
+    from langchain.embeddings.openai import OpenAIEmbeddings
+
+    embedder = OpenAIEmbeddings(
+        openai_api_key=api_key,
+        model=model_name,
+    )
+
+    _ = embedder.embed_query("We are testing authentication")
+    return embedder
+
+
+@dataclass
+class OpenAIEmbedderConfig(ConfigSessionHandleMixin):
+    def __init__(
+        self,
+        api_key: str,
+        list_of_elements_json_paths: List[str],
+        output_dir: str,
+        model_name: Optional[str] = None,
+    ):
+        self.api_key = api_key
+        self.model_name = model_name if model_name is not None else "text-embedding-ada-002"
+        self.list_of_elements_json_paths = list_of_elements_json_paths
+        self.output_dir = output_dir
+
+    def create_session_handle(
+        self,
+    ) -> OpenAISessionHandle:
+        service = create_openai_object(self.api_key, self.model_name)
+        return OpenAISessionHandle(service=service)
+
+
+class ElementEncoder(json.JSONEncoder):
+    def default(self, obj):
+        if isinstance(obj, Element):
+            return {
+                "text": obj.text,
+                "element_id": obj.id,
+                "embedding": obj.embedding,
+                "metadata": obj.metadata.to_dict(),
+            }
+        return super().default(obj)
+
+
+@dataclass
+class OpenAIEmbeddingsDoc(IngestDocSessionHandleMixin, DataClassJsonMixin):
+    """Class encapsulating reading elements in a doc and writing the obtained embeddings."""
+
+    config: OpenAIEmbedderConfig
+    filename: str
+
+    @property
+    def _output_filename(self):
+        """Create output file path."""
+        output_file = f"{self.filename}_with_embeddings.json"
+
+        return (Path(self.config.output_dir) / output_file).resolve()
+
+    def get_elements_from_json(self) -> List[Element]:
+        """Reads elements"""
+        return partition_json(filename=self.filename)
+
+    @requires_dependencies(["langchain"], extras="jira")
+    def embed_and_write_result(self):
+        logger.debug(f"Reading: {self} - PID: {os.getpid()}")
+
+        elements = self.get_elements_from_json()
+        embeddings = self.session_handle.service.embed_documents([str(e) for e in elements])
+        self.elements = self.add_embeddings_to_elements(elements, embeddings)
+        # import pdb; pdb.set_trace()
+        self.write_result()
+
+    def add_embeddings_to_elements(self, elements, embeddings):
+        assert len(elements) == len(embeddings)
+
+        for i in range(len(elements)):
+            elements[i].embedding = embeddings[i]
+
+        return elements
+
+    def write_result(self):
+        """Write the structured json result for this doc. result must be json serializable."""
+        self._output_filename.parent.mkdir(parents=True, exist_ok=True)
+        with open(self._output_filename, "w", encoding="utf8") as output_f:
+            json.dump(self.elements, output_f, ensure_ascii=False, indent=2, cls=ElementEncoder)
```

Comment:
we are probably just prototyping here so writing to json is fine but depending on the embedding model this can blow up the json file size very quickly. e.g., the most commonly used ones have 748 dimension, so that is 748 floating point numbers per embedding.
speaking of which we should probably record encoder metadata somewhere in the process/output out well, like the vector dimension and if it is unit vector (two attributes relevant to information retrieval tasks -> computing distance)

Reply:
On saving the embeddings to the disk, what are some good alternatives? 

What I think is:
- If we'd like to isolate the task of encoding from the task of writing to a DB, and want an intermediate place to store the embeddings:
  - we can save embeddings to a separate file, again in a basic format. these files can be processed later on to be written into a database
  - we can use a message queue so the embeddings later get consumed by a job and be written into the database
- Or, we could directly write the embeddings into a database

Reply:
do we actually need to write this to file? or can we just rely on return the in-memory elements with embeddings?

Reply:
Depends on how the downstream jobs want to obtain and use these embeddings. If we'd like embed large corpora, we'd like to index the embeddings in a DB. To do that, the job that'll write the embeddings into the DB can read the embeddings from memory, from a message queue, or from the disk

If adequate, we can return the embeddings in-memory for now; and open another ticket to bring in persistence

Reply:
> Depends on how the downstream jobs want to obtain and use these embeddings
yes, definitely. lets start with the requirements for ingest and what makes most sense there and how this interfaces.  CC @rbiseck3 

aside: let's avoid a message queue

Reply:
Update: For the current ticket, we've decided to make this work fully in memory. Inputs are read from memory, results are passed on via memory. https://github.com/Unstructured-IO/unstructured/pull/1368/commits/499795ae9364593a4954a30e0e568598f3d15f33